kubetest2/pkg/metadata/junit.go

```diff
@@ -0,0 +1,73 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metadata
+
+import (
+	"encoding/xml"
+	"io"
+)
+
+// JUnitError represents an error with extra metadata suitable for structured
+// (JUnit.xml) output written by kubetest2
+// If this type is returned as an error by a Deployer, kubetest2 may write the
+// metadata to the output junit_runner.xml
+type JUnitError interface {
+	error
+	Stdout() string
+	Stderr() string
+}
+
+// testSuite holds a slice of TestCase and other summary metadata.
+//
+// A build (column in testgrid) is composed of one or more TestSuites.
+type testSuite struct {
+	XMLName  xml.Name `xml:"testsuite"`
+	Failures int      `xml:"failures,attr"`
+	Tests    int      `xml:"tests,attr"`
+	Time     float64  `xml:"time,attr"`
+	Cases    []testCase
+}
+
+func (t *testSuite) Write(writer io.Writer) error {
+	// write xml header
+	io.WriteString(writer, `<?xml version="1.0" encoding="UTF-8"?>`)
+	// write indented suite
+	e := xml.NewEncoder(writer)
+	e.Indent("", "   ")
```

Comment:
Again, this is more or less what kubetest does (which is correct), but we use an encoder to write directly to the writer instead of marshalling and then writing the bytes to the writer.

https://github.com/kubernetes/test-infra/blob/3c776246ad7f2c801d41626b43069e10ddefcf91/kubetest/process/process.go#L65-L97

Reply:
why indent with 3 spaces instead of an even number?

Reply:
good question, I started with matching kubetest v1 but I think a space got dropped there

Reply:
this is 4 indent now